File: vsub/subtitle.py
"""字幕生成模块"""

import re
from pathlib import Path
from typing import List

from vsub.asr import AsrWord
from vsub.config import OutputFormat


class SubtitleSegment:
    """字幕片段"""

    def __init__(
        self,
        index: int,
        start: float,
        end: float,
        text: str,
    ):
        self.index = index
        self.start = start
        self.end = end
        self.text = text

    def __repr__(self) -> str:
        return f"SubtitleSegment({self.index}, {self.start:.2f}-{self.end:.2f}, {self.text!r})"

    @staticmethod
    def format_time(seconds: float, separator: str = ",") -> str:
        """格式化时间戳为 SRT/VTT 格式 (HH:MM:SS{separator}mmm)"""
        hours = int(seconds // 3600)
        minutes = int((seconds % 3600) // 60)
        secs = int(seconds % 60)
        millis = int((seconds % 1) * 1000)
        return f"{hours:02d}:{minutes:02d}:{secs:02d}{separator}{millis:03d}"

    @staticmethod
    def format_time_srt(seconds: float) -> str:
        """格式化时间戳为 SRT 格式 (HH:MM:SS,mmm)"""
        return SubtitleSegment.format_time(seconds, separator=",")

    @staticmethod
    def format_time_vtt(seconds: float) -> str:
        """格式化时间戳为 VTT 格式 (HH:MM:SS.mmm)"""
        return SubtitleSegment.format_time(seconds, separator=".")


class SubtitleGenerator:
    """字幕生成器"""

    def __init__(self, max_line_length: int = 80, max_line_count: int = 2):
        self.max_line_length = max_line_length
        self.max_line_count = max_line_count

    def generate_srt(self, segments: List[SubtitleSegment]) -> str:
        """生成 SRT 格式字幕"""
        lines = []
        for segment in segments:
            lines.append(str(segment.index))
            start = SubtitleSegment.format_time_srt(segment.start)
            end = SubtitleSegment.format_time_srt(segment.end)
            lines.append(f"{start} --> {end}")
            lines.append(self._wrap_text(segment.text))
            lines.append("")  # 空行分隔
        return "\n".join(lines)

    def generate_vtt(self, segments: List[SubtitleSegment]) -> str:
        """生成 VTT 格式字幕"""
        lines = ["WEBVTT", ""]
        for segment in segments:
            start = SubtitleSegment.format_time_vtt(segment.start)
            end = SubtitleSegment.format_time_vtt(segment.end)
            lines.append(f"{start} --> {end}")
            lines.append(self._wrap_text(segment.text))
            lines.append("")  # 空行分隔
        return "\n".join(lines)

    def generate(self, segments: List[SubtitleSegment], fmt: OutputFormat) -> str:
        """根据格式生成字幕"""
        if fmt == OutputFormat.SRT:
            return self.generate_srt(segments)
        elif fmt == OutputFormat.VTT:
            return self.generate_vtt(segments)
        else:
            raise ValueError(f"不支持的格式: {fmt}")

    def _wrap_text(self, text: str) -> str:
        """文本自动换行"""
        words = text.split()
        if not words:
            return ""

        lines = []
        current_line = []
        current_length = 0

        for word in words:
            word_len = len(word)
            # 检查当前行加上这个单词是否超过最大长度
            new_length = current_length + (1 if current_line else 0) + word_len

            if current_line and new_length > self.max_line_length:
                # 当前行已满，保存并开始新行
                lines.append(" ".join(current_line))
                current_line = [word]
                current_length = word_len
            else:
                # 添加到当前行
                current_line.append(word)
                current_length = new_length

            # 检查是否达到最大行数限制
            if len(lines) >= self.max_line_count:
                # 保存当前行，清空以继续处理剩余单词
                if current_line:
                    lines.append(" ".join(current_line))
                    current_line = []
                    current_length = 0
                break

        # 添加最后一行
        if current_line:
            lines.append(" ".join(current_line))

        return "\n".join(lines[:self.max_line_count])

    def write_to_file(
        self,
        segments: List[SubtitleSegment],
        fmt: OutputFormat,
        path: Path,
        overwrite: bool = False,
    ) -> None:
        """写入字幕文件"""
        if path.exists() and not overwrite:
            raise FileExistsError(f"文件已存在（使用 --overwrite 覆盖）: {path}")

        content = self.generate(segments, fmt)
        path.write_text(content, encoding="utf-8")


def segments_from_asr_result(
    words: List[AsrWord],
    max_duration: float = 5.0,
    max_chars: int = 80,
) -> List[SubtitleSegment]:
    """从 ASR 结果创建字幕片段"""
    segments = []
    current_words = []
    segment_start = 0.0
    char_count = 0

    # 句子结束标点（修复：检测单词末尾包含标点）
    sentence_end = re.compile(r"[.!?。！？]")

    for i, word in enumerate(words):
        if not current_words:
            segment_start = word.start

        current_words.append(word)
        # 修复：字符计数考虑空格
        char_count += len(word.text)
        if current_words:  # 不是第一个单词，添加空格
            char_count += 1

        current_duration = word.end - segment_start
        is_last = i == len(words) - 1

        # 检查是否需要分割
        # 修复：检测单词末尾是否包含句子结束标点
        has_end_punctuation = bool(sentence_end.search(word.text)) if word.text else False
        should_split = (
            is_last
            or current_duration >= max_duration
            or char_count >= max_chars
            or has_end_punctuation
        )

        if should_split and current_words:
            segment_text = " ".join(w.text for w in current_words)
            segment_end = current_words[-1].end

            segments.append(
                SubtitleSegment(
                    index=len(segments) + 1,
                    start=segment_start,
                    end=segment_end,
                    text=segment_text,
                )
            )

            current_words = []
            char_count = 0

    return segments

File: src/core/subtitle.rs
use std::fmt::Write;

use crate::core::config::OutputFormat;
use crate::asr::AsrWord;
use crate::VsubError;

/// 字幕片段
#[derive(Debug, Clone)]
pub struct SubtitleSegment {
    pub index: usize,
    pub start: f64,
    pub end: f64,
    pub text: String,
}

impl SubtitleSegment {
    pub fn new(index: usize, start: f64, end: f64, text: String) -> Self {
        Self { index, start, end, text }
    }

    /// 格式化时间戳为 SRT 格式 (HH:MM:SS,mmm)
    pub fn format_time_srt(seconds: f64) -> String {
        let hours = (seconds / 3600.0) as u32;
        let minutes = ((seconds % 3600.0) / 60.0) as u32;
        let secs = (seconds % 60.0) as u32;
        let millis = ((seconds % 1.0) * 1000.0) as u32;

        format!("{:02}:{:02}:{:02},{:03}", hours, minutes, secs, millis)
    }

    /// 格式化时间戳为 VTT 格式 (HH:MM:SS.mmm)
    pub fn format_time_vtt(seconds: f64) -> String {
        let hours = (seconds / 3600.0) as u32;
        let minutes = ((seconds % 3600.0) / 60.0) as u32;
        let secs = (seconds % 60.0) as u32;
        let millis = ((seconds % 1.0) * 1000.0) as u32;

        format!("{:02}:{:02}:{:02}.{:03}", hours, minutes, secs, millis)
    }
}

/// 字幕生成器
pub struct SubtitleGenerator {
    max_line_length: usize,
    max_line_count: usize,
}

impl SubtitleGenerator {
    pub fn new(max_line_length: usize, max_line_count: usize) -> Self {
        Self {
            max_line_length,
            max_line_count,
        }
    }

    /// 生成 SRT 格式字幕
    pub fn generate_srt(&self, segments: &[SubtitleSegment]) -> Result<String, VsubError> {
        let mut output = String::new();

        for segment in segments {
            // 序号
            writeln!(&mut output, "{}", segment.index)
                .map_err(|e| VsubError::Subtitle(format!("写入失败: {}", e)))?;

            // 时间戳
            let start = SubtitleSegment::format_time_srt(segment.start);
            let end = SubtitleSegment::format_time_srt(segment.end);
            writeln!(&mut output, "{} --> {}", start, end)
                .map_err(|e| VsubError::Subtitle(format!("写入失败: {}", e)))?;

            // 文本（自动换行）
            let wrapped = self.wrap_text(&segment.text);
            writeln!(&mut output, "{}", wrapped)
                .map_err(|e| VsubError::Subtitle(format!("写入失败: {}", e)))?;

            // 空行分隔
            writeln!(&mut output)
                .map_err(|e| VsubError::Subtitle(format!("写入失败: {}", e)))?;
        }

        Ok(output)
    }

    /// 生成 VTT 格式字幕
    pub fn generate_vtt(&self, segments: &[SubtitleSegment]) -> Result<String, VsubError> {
        let mut output = String::new();

        // VTT 头部
        writeln!(&mut output, "WEBVTT")
            .map_err(|e| VsubError::Subtitle(format!("写入失败: {}", e)))?;
        writeln!(&mut output)
            .map_err(|e| VsubError::Subtitle(format!("写入失败: {}", e)))?;

        for segment in segments {
            // 时间戳（VTT 不需要序号）
            let start = SubtitleSegment::format_time_vtt(segment.start);
            let end = SubtitleSegment::format_time_vtt(segment.end);
            writeln!(&mut output, "{} --> {}", start, end)
                .map_err(|e| VsubError::Subtitle(format!("写入失败: {}", e)))?;

            // 文本（自动换行）
            let wrapped = self.wrap_text(&segment.text);
            writeln!(&mut output, "{}", wrapped)
                .map_err(|e| VsubError::Subtitle(format!("写入失败: {}", e)))?;

            // 空行分隔
            writeln!(&mut output)
                .map_err(|e| VsubError::Subtitle(format!("写入失败: {}", e)))?;
        }

        Ok(output)
    }

    /// 根据格式生成字幕
    pub fn generate(
        &self,
        segments: &[SubtitleSegment],
        format: OutputFormat,
    ) -> Result<String, VsubError> {
        match format {
            OutputFormat::Srt => self.generate_srt(segments),
            OutputFormat::Vtt => self.generate_vtt(segments),
        }
    }

    /// 文本自动换行
    fn wrap_text(&self, text: &str) -> String {
        let words: Vec<&str> = text.split_whitespace().collect();
        let mut lines: Vec<String> = Vec::new();
        let mut current_line = String::new();

        for word in words {
            // 检查是否需要换行
            if !current_line.is_empty()
                && current_line.len() + 1 + word.len() > self.max_line_length
            {
                lines.push(current_line);
                current_line = String::new();

                // 检查是否达到最大行数
                if lines.len() >= self.max_line_count {
                    // 剩余词作为新行开始
                    current_line = word.to_string();
                    continue;
                }
            }

            if !current_line.is_empty() {
                current_line.push(' ');
            }
            current_line.push_str(word);
        }

        if !current_line.is_empty() {
            lines.push(current_line);
        }

        lines.join("\n")
    }

    /// 写入字幕文件
    pub fn write_to_file<P: AsRef<std::path::Path>>(
        &self,
        segments: &[SubtitleSegment],
        format: OutputFormat,
        path: P,
        overwrite: bool,
    ) -> Result<(), VsubError> {
        let path = path.as_ref();

        // 检查文件是否已存在
        if path.exists() && !overwrite {
            return Err(VsubError::Subtitle(format!(
                "文件已存在（使用 --overwrite 覆盖）: {}",
                path.display()
            )));
        }

        let content = self.generate(segments, format)?;

        std::fs::write(path, content)
            .map_err(|e| VsubError::Io(e))?;

        Ok(())
    }
}

/// 从 ASR 结果创建字幕片段
pub fn segments_from_asr_result(
    words: &[AsrWord],
    max_duration: f64,
    max_chars: usize,
) -> Vec<SubtitleSegment> {
    let mut segments: Vec<SubtitleSegment> = Vec::new();
    let mut current_words: Vec<&AsrWord> = Vec::new();
    let mut segment_start: f64 = 0.0;
    let mut char_count: usize = 0;

    for (i, word) in words.iter().enumerate() {
        // 开始新片段
        if current_words.is_empty() {
            segment_start = word.start;
        }

        current_words.push(word);
        char_count += word.text.len();

        let current_duration = word.end - segment_start;
        let is_last = i == words.len() - 1;

        // 检查是否需要分割
        let should_split = is_last
            || current_duration >= max_duration
            || char_count >= max_chars
            || word.text.ends_with(['.', '!', '?', '。', '！', '？']);

        if should_split && !current_words.is_empty() {
            let segment_text = current_words
                .iter()
                .map(|w| w.text.as_str())
                .collect::<Vec<_>>()
                .join(" ");

            let segment_end = current_words.last().unwrap().end;

            segments.push(SubtitleSegment::new(
                segments.len() + 1,
                segment_start,
                segment_end,
                segment_text,
            ));

            current_words.clear();
            char_count = 0;
        }
    }

    segments
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_time_srt() {
        assert_eq!(SubtitleSegment::format_time_srt(0.0), "00:00:00,000");
        assert_eq!(SubtitleSegment::format_time_srt(90.5), "00:01:30,500");
        assert_eq!(SubtitleSegment::format_time_srt(3661.123), "01:01:01,123");
    }

    #[test]
    fn test_format_time_vtt() {
        assert_eq!(SubtitleSegment::format_time_vtt(0.0), "00:00:00.000");
        assert_eq!(SubtitleSegment::format_time_vtt(90.5), "00:01:30.500");
    }

    #[test]
    fn test_wrap_text() {
        let generator = SubtitleGenerator::new(20, 2);
        let text = "This is a very long text that should be wrapped properly";
        let wrapped = generator.wrap_text(text);
        let lines: Vec<&str> = wrapped.lines().collect();
        assert!(lines.len() <= 2);
        assert!(lines.iter().all(|line| line.len() <= 20));
    }

    #[test]
    fn test_generate_srt() {
        let generator = SubtitleGenerator::new(40, 2);
        let segments = vec![
            SubtitleSegment::new(1, 0.0, 3.5, "Hello world".to_string()),
            SubtitleSegment::new(2, 3.5, 7.0, "Second subtitle".to_string()),
        ];

        let srt = generator.generate_srt(&segments).unwrap();
        assert!(srt.contains("1"));
        assert!(srt.contains("00:00:00,000 --> 00:00:03,500"));
        assert!(srt.contains("Hello world"));
    }

    #[test]
    fn test_segments_from_asr_result() {
        let words = vec![
            AsrWord { text: "Hello".to_string(), start: 0.0, end: 0.5, confidence: 0.9 },
            AsrWord { text: "world".to_string(), start: 0.6, end: 1.0, confidence: 0.95 },
            AsrWord { text: ".".to_string(), start: 1.0, end: 1.1, confidence: 1.0 },
        ];

        let segments = segments_from_asr_result(&words, 5.0, 50);
        assert_eq!(segments.len(), 1);
        assert_eq!(segments[0].text, "Hello world .");
    }
}
